plugins/inputs/mavlink/mavlink.go

```diff
@@ -0,0 +1,239 @@
+	return sampleConfig
+}
+
+func (m *Mavlink) Init() error {
+	// Set default values
+	if m.URL == "" {
+		m.URL = "tcp://127.0.0.1:5760"
+	}
+
+	// Parse out the Mavlink endpoint.
+	// Try to parse the URL
+	u, err := url.Parse(m.URL)
+	if err != nil {
+		return fmt.Errorf("invalid url: %w", err)
+	}
+
+	switch u.Scheme {
+	case "serial":
+		// Serial client
+		// Parse serial URL by hand, because it is not a compliant URL.
+		baudRate := 57600
+		device, rate, found := strings.Cut(strings.TrimPrefix(m.URL, "serial://"), ":")
+		if found {
+			r, err := strconv.Atoi(rate)
+			if err != nil {
+				return fmt.Errorf("serial baud rate not valid: %w", err)
+			}
+			baudRate = r
+		}
+
+		m.endpointConfig = gomavlib.EndpointSerial{
+			Device: device,
+			Baud:   baudRate,
+		}
+	case "tcp":
+		// Use default TCP port if it was not specified
+		port := u.Port()
+		if port == "" {
+			port = "5760"
+		}
+
+		if u.Hostname() == "" {
+			m.endpointConfig = gomavlib.EndpointTCPServer{
+				Address: "0.0.0.0:" + port,
+			}
+		} else {
+			m.endpointConfig = gomavlib.EndpointTCPClient{
+				Address: net.JoinHostPort(u.Hostname(), port),
+			}
+		}
+
+	case "udp":
+		// Use default UDP port if it was not specified
+		port := u.Port()
+		if port == "" {
+			port = "14550"
+		}
+
+		if u.Hostname() == "" {
+			m.endpointConfig = gomavlib.EndpointUDPServer{
+				Address: "0.0.0.0:" + port,
+			}
+		} else {
+			m.endpointConfig = gomavlib.EndpointUDPClient{
+				Address: net.JoinHostPort(u.Hostname(), port),
+			}
+		}
+
+	default:
+		return fmt.Errorf("unknown scheme %q", u.Scheme)
+	}
+
+	// Compile filter
+	m.filter, err = filter.Compile(m.FilterPattern)
+	if err != nil {
+		return fmt.Errorf("compiling filter failed: %w", err)
+	}
+
+	return nil
+}
+
+func (m *Mavlink) Start(acc telegraf.Accumulator) error {
+	// Start MAVLink endpoint
+	connection, err := gomavlib.NewNode(gomavlib.NodeConf{
+		Endpoints:              []gomavlib.EndpointConf{m.endpointConfig},
+		Dialect:                ardupilotmega.Dialect,
+		OutVersion:             gomavlib.V2,
+		OutSystemID:            m.SystemID,
+		StreamRequestEnable:    m.StreamRequestFrequency > 0,
+		StreamRequestFrequency: int(m.StreamRequestFrequency),
+	})
+	if err != nil {
+		return &internal.StartupError{
+			Err:   fmt.Errorf("connecting to mavlink endpoint %s failed: %w", m.URL, err),
+			Retry: true,
+		}
+	}
+	m.connection = connection
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	m.cancel = cancelFunc
+
+	// Start routine to connect to Mavlink and stream out data async
+	m.wg.Add(1)
+	go func(ctx context.Context) {
+		defer m.connection.Close()
+		defer m.wg.Done()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case evt := <-m.connection.Events():
+				switch evt := evt.(type) {
+				case *gomavlib.EventFrame:
+					if result := convertFrameToMetric(evt.Frame, m.filter); result != nil {
+						result.AddTag("source", m.URL)
+						acc.AddMetric(result)
+					}
```

Comment:
How about making `convertFrameToMetric` a member of the plugin and pass in the accumulator? This way the handling can fully happen in that function
```suggestion
					m.handleFrame(acc, evt.Frame)
```

Reply:
Makes sense, updated to restructure things this way.